openml/datasets/dataset.py

```diff
@@ -290,30 +353,98 @@ def decode_arff(fh):
             with io.open(filename, encoding='utf8') as fh:
                 return decode_arff(fh)
 
+    @staticmethod
+    def _convert_array_format(data, array_format, attribute_names):
+        """Convert a dataset to a given array format.
+
+        By default, the data are stored as a sparse matrix or a pandas
+        dataframe. One might be interested to get a pandas SparseDataFrame or a
+        NumPy array instead, respectively.
+        """
+        if array_format == "array" and not scipy.sparse.issparse(data):
+            try:
+                # for back-compatibility, we try try to convert the data to
+                # float 32 bits.
+                return np.asarray(data, dtype=np.float32)
+            except ValueError:
+                return np.asarray(data)
+        if array_format == "dataframe" and scipy.sparse.issparse(data):
+            return pd.SparseDataFrame(data, columns=attribute_names)
+        return data
+
+    @staticmethod
+    def _create_column_int_with_NA(series):
+        """Convert to object an int series containing missing values."""
+        col = [np.nan
+               if x is None or np.isnan(x)
+               else int(x) for x in series]
+        return pd.Series(col, index=series.index, dtype=object)
+
+    @staticmethod
+    def _create_column_bool(series, categories):
+        col = []
+        for x in series:
+            try:
+                val = categories[int(x)]
+                val = True if val.lower() == 'true' else False
+                col.append(val)
+            except (TypeError, ValueError):
+                col.append(np.nan)
+        return pd.Series(col, index=series.index, dtype=object)
```

Comment:
Is there a reason to keep this a type object and not cast to categorical?

Reply:
There is no need to force the dtype here. It will be `bool` with there is not missing values and `object` if there is. `object` refers to `bool` and `float` (i.e., np.nan marker).

Reply:
But in the case the column is all bool, there should be only two allowed values in total. Therefore, the data frame should reflect that for two reasons:
1. to have all information about the dataset in the dataframe (object is ambiguous here)
2. to not allow adding any other types to that column

Reply:
My previous comment was not clear enough. When I meant `There is no need to force the dtype here`, I mean that I should remove `dtype=object` because the type inference will work. In the case of pure `bool` the type will be `bool`.

Reply:
Okay, got it. I just looked up the specification of arff again, and there is no definition of bool. Could you please state your opinion on casting boolean columns to categorical?

Reply:
My only concern is that the round trip (upload-download) will not give the original data. If you are fine with that then we can just let the bool as categories and it will be encoded by any encoder after anyway.

Reply:
Hm, why would a round-trip destroy categorical-encoded bools? If a category with `['True', 'False']` is passed it should be serialized/deserialized without any loss in information.

Reply:
What I meant is that a user will pass a dataframe with `[True, False]` in a column with `dtype=bool` and will get back a `['True', 'False']` column with `dtype='categories'`.

Reply:
So the dtype and the values are not the same even if they represnet the same information. Anyway, I would say that this is also a limitation of the ARFF format not having a proper dtype info.

Reply:
Got it, thanks for the explanation. I'd be fine with that behavior, given that this comes from the arff format not defining boolean data types.